File: bcw5/php/Mohican/Mohican.php
<?php
	class Mohican {
		static $lastMohican;
	    static $id = 0;
	    private $selfID;
	    private $name;

	    public function Mohican($name) {
	        $this->selfID = ++self::$id;
	        $this->name = $name;
	        self::$lastMohican = $this;
	    }

	    public function __clone() {
	        $this->selfID = ++self::$id;
	        $this->name = $this->name;
	        self::$lastMohican = $this;
	    }

	    public function setName($name) {
	    	$this->name = $name;
	    }
	}
?>

File: bcw4/Java_Bcw4/Mohican/Mohican.java
public class Mohican {

    private static Mohican lastMohican;
    private static int id;
    private int selfID;
    private String name;

    public Mohican(String name) {
        Mohican.id += 1;
        selfID = Mohican.id;
        this.name = name;
        Mohican.lastMohican = this;
    }

    public int getSelfID() {
        return selfID;
    }

    public String getName() {
        return name;
    }

    public static int getId() {
        return id;
    }

    public static Mohican getLast() {
        return lastMohican;
    }

    @Override
    public String toString() {
        return "Mohican" + "(" + name + ", " + selfID + ")";
    }

    @Override
    protected void finalize() throws Throwable {
        System.out.println("Passport " + name +" destroyed.");
    }
}
